infrastructures/docker/environment/src/main/java/org/eclipse/che/workspace/infrastructure/docker/environment/compose/ComposeEnvironmentFactory.java

```diff
@@ -0,0 +1,113 @@
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.dataformat.yaml.YAMLFactory;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableSet;
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import org.eclipse.che.api.core.ValidationException;
+import org.eclipse.che.api.core.model.workspace.Warning;
+import org.eclipse.che.api.installer.server.InstallerRegistry;
+import org.eclipse.che.api.workspace.server.spi.InfrastructureException;
+import org.eclipse.che.api.workspace.server.spi.environment.InternalEnvironmentFactory;
+import org.eclipse.che.api.workspace.server.spi.environment.InternalMachineConfig;
+import org.eclipse.che.api.workspace.server.spi.environment.InternalRecipe;
+import org.eclipse.che.api.workspace.server.spi.environment.RecipeRetriever;
+import org.eclipse.che.workspace.infrastructure.docker.environment.compose.model.ComposeRecipe;
+
+/** @author Sergii Leshchenko */
+@Singleton
+public class ComposeEnvironmentFactory extends InternalEnvironmentFactory<ComposeEnvironment> {
+
+  private static final ObjectMapper YAML_PARSER = new ObjectMapper(new YAMLFactory());
+
+  private static Set<String> YAML_CONTENT_TYPES =
+      ImmutableSet.of("application/x-yaml", "text/yaml", "text/x-yaml");
+
+  private final ComposeServicesStartStrategy startStrategy;
+
+  @Inject
+  public ComposeEnvironmentFactory(
+      InstallerRegistry installerRegistry,
+      RecipeRetriever recipeRetriever,
+      ComposeEnvironmentValidator composeValidator,
+      ComposeServicesStartStrategy startStrategy) {
+    super(installerRegistry, recipeRetriever, composeValidator);
+    this.startStrategy = startStrategy;
+  }
+
+  @Override
+  protected ComposeEnvironment doCreate(
+      InternalRecipe recipe, Map<String, InternalMachineConfig> machines, List<Warning> warnings)
+      throws InfrastructureException, ValidationException {
+    String contentType = recipe.getContentType();
+    checkNotNull(contentType, "Recipe content type should not be null");
+
+    String recipeContent = recipe.getContent();
+
+    if (!ComposeEnvironment.TYPE.equals(recipe.getType())) {
+      throw new ValidationException(
+          format("Compose environment parser doesn't support recipe type '%s'", recipe.getType()));
+    }
+
+    if (!YAML_CONTENT_TYPES.contains(contentType)) {
+      throw new ValidationException(
+          format(
+              "Provided environment recipe content type '%s' is "
+                  + "unsupported. Supported values are: %s",
+              contentType, YAML_CONTENT_TYPES.stream().collect(joining(", "))));
+    }
+    ComposeRecipe composeRecipe = doParse(recipeContent);
+
+    // check that containers start order can be resolved
+    // NOTE: it should be performed before dockerEnvironmentNormalizer.normalize because
```

Comment:
Looks like this comment is irrelevant now. WDYT?

Reply:
You're right. I'll remove this comment

Reply:
Removed